File: akka-scala-vs-java/scala/stash/src/main/scala/com/github/daggerok/akka/stash/AkkaStash.scala
package com.github.daggerok.akka.stash

import akka.actor.{Actor, ActorLogging, ActorRef, ActorSystem, Props, Stash}

case class User(name: String, age: Int)

object UserRepositoryActor {
  sealed trait Messages
  case object Connect extends Messages
  case object Disconnect extends Messages

  sealed trait DbOperations extends Messages
  case class Save(user: User) extends DbOperations
  case class Delete(user: User) extends DbOperations
  //etc...
}

class UserRepositoryActor extends Actor with ActorLogging with Stash {
  import UserRepositoryActor._

  override def receive: Receive = disconnected

  def disconnected: Receive = {
    case Connect =>
      log.info("connected!")
      unstashAll()
      context.become(connected)
    case _ =>
      log.info("received a message when disconnected!")
      stash()
  }

  def connected: Receive = {
    case Disconnect =>
      log.info("disconnect!")
      context.unbecome()
    case op: DbOperations =>
      log.info("received: {}", op)
  }
}

object AkkaStash extends App {
  val system: ActorSystem = ActorSystem("stashAkka")
  val userRepositoryActor = system.actorOf(Props[UserRepositoryActor], "userRepositoryActor")

  import UserRepositoryActor._
  userRepositoryActor ! Save(User("max", 123))
  userRepositoryActor ! Connect
  userRepositoryActor ! Save(User("bob", 321))
  userRepositoryActor ! Delete(User("bob", 321))
  userRepositoryActor ! Disconnect

  //Thread.sleep(123)
  system.terminate()
}

File: akka-scala-vs-java/java/persistence/counter/src/main/java/com/github/daggerok/akka/stash/AkkaStash.java
package com.github.daggerok.akka.stash;

import akka.actor.AbstractActorWithStash;
import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import lombok.Value;
import lombok.extern.slf4j.Slf4j;

import java.io.Serializable;

@Value(staticConstructor = "of")
final class User implements Serializable {
    private final String name;
    private final int age;
}

interface Messages extends Serializable { }
final class Connect implements Messages { }
final class Disconnect implements Messages { }

interface DbOperations extends Messages { }

@Value(staticConstructor = "of")
final class Save implements DbOperations {
    private final User user;
}

@Value(staticConstructor = "of")
final class Delete implements DbOperations {
    private final User user;
}

// etc...

@Slf4j
class UserRepositoryActor extends AbstractActorWithStash {

    static Props props() {
        return Props.create(UserRepositoryActor.class);
    }

    @Override
    public Receive createReceive() {
        return disconnected();
    }

    private Receive disconnected() {
        return receiveBuilder()
                .match(Connect.class, connect -> {
                    log.info("connected!");
                    unstashAll();
                    getContext().become(connected());
                })
                .matchAny(o -> {
                    log.info("received a message when disconnected!");
                    stash();
                })
                .build();
    }

    private Receive connected() {
        return receiveBuilder()
                .match(Disconnect.class, disconnect -> {
                    log.info("disconnect!");
                    getContext().unbecome();
                })
                .match(DbOperations.class, op -> {
                    log.info("received: {}", op);
                })
                .matchAny(o -> {
                    log.error("received unknown: {}", o);
                    unhandled(o);
                })
                .build();
    }
}

public class AkkaStash {
    public static void main(String[] args) {
        ActorSystem system = ActorSystem.create("stashAkka");
        ActorRef userRepositoryActor = system.actorOf(UserRepositoryActor.props(), "userRepositoryActor");
        ActorRef sender = ActorRef.noSender();

        userRepositoryActor.tell(Save.of(User.of("max", 123)), sender);
        userRepositoryActor.tell(new Connect(), sender);
        userRepositoryActor.tell(Save.of(User.of("bob", 321)), sender);
        userRepositoryActor.tell(Delete.of(User.of("bob", 321)), sender);
        userRepositoryActor.tell(new Disconnect(), sender);

        system.terminate();
    }
}
